Sources/NIOHTTP1/HTTPUpgradeHandler.swift

```diff
@@ -131,45 +131,57 @@ public class HTTPServerUpgradeHandler: ChannelInboundHandler {
     }
 
     public func channelRead(ctx: ChannelHandlerContext, data: NIOAny) {
-        if self.upgrading {
+        guard !self.seenFirstRequest else {
             // We're waiting for upgrade to complete: buffer this data.
             self.receivedMessages.append(data)
             return
         }
 
-        // We're trying to remove ourselves from the pipeline but not upgrading, so just pass this on.
-        if seenFirstRequest {
-            ctx.fireChannelRead(data)
-            return
-        }
-
         let requestPart = unwrapInboundIn(data)
-        seenFirstRequest = true
-
-        // We should only ever see a request header: by the time the body comes in we should
-        // be out of the pipeline. Anything else is an error.
-        guard case .head(let request) = requestPart else {
-            ctx.fireErrorCaught(HTTPUpgradeErrors.invalidHTTPOrdering)
-            notUpgrading(ctx: ctx, data: data)
-            return
-        }
 
-        // Ok, we have a HTTP request. Check if it's an upgrade. If it's not, we want to pass it on and remove ourselves
-        // from the channel pipeline.
-        let requestedProtocols = request.headers[canonicalForm: "upgrade"]
-        guard requestedProtocols.count > 0 else {
-            notUpgrading(ctx: ctx, data: data)
-            return
-        }
-
-        // Cool, this is an upgrade! Let's go.
-        if !handleUpgrade(ctx: ctx, request: request, requestedProtocols: requestedProtocols) {
-            notUpgrading(ctx: ctx, data: data)
+        if let upgrade = self.upgrade {
+            switch requestPart {
+            case .head(_):
+                ctx.fireErrorCaught(HTTPUpgradeErrors.invalidHTTPOrdering)
+                notUpgrading(ctx: ctx, data: data)
+                return
+            case .body(_):
+                // ignore
```

Comment:
Can you mark this with a TODO that indicates that in a future NIO version we want to add an API where we deliver this to the upgrader in some form?